src/OrleansRuntime/GrainDirectory/RemoteGrainDirectory.cs

```diff
@@ -336,24 +75,24 @@ public Task RegisterManySingleActivation(List<ActivationAddress> addresses, int
                 if (curGen == tuple.Item2 || curGen == GrainInfo.NO_ETAG)
                 {
                     // the grain entry either does not exist in the local partition (curGen = -1) or has not been updated
-                    result.Add(new Tuple<GrainId, int, List<Tuple<SiloAddress, ActivationId>>>(tuple.Item1, curGen, null));
+                    result.Add(new Tuple<GrainId, int, List<ActivationAddress>>(tuple.Item1, curGen, null));
                 }
                 else
                 {
                     // the grain entry has been updated -- fetch and return its current version
-                    Tuple<List<Tuple<SiloAddress, ActivationId>>, int> lookupResult = partition.LookUpGrain(tuple.Item1);
+                    var lookupResult = await router.LookupAsync(tuple.Item1, 1);
```

Comment:
Now that we are awaiting for a potential remote lookup, should the foreach start the tasks and then do a single await Task.WhenAll of all the lookups? BTW, was the change to look up in the local partition to do potential hops by design in this method? Maybe @gabikliot has input for this change?